database/pebble/db.go

```diff
@@ -0,0 +1,326 @@
+	"github.com/ava-labs/avalanchego/utils/set"
+	"github.com/ava-labs/avalanchego/utils/units"
+)
+
+// pebbleByteOverHead is the number of bytes of constant overhead that
+// should be added to a batch size per operation.
+const pebbleByteOverHead = 8
+
+var (
+	_ database.Database = (*Database)(nil)
+
+	ErrInvalidOperation = errors.New("invalid operation")
+
+	defaultCacheSize = 512 * units.MiB
+	DefaultConfig    = Config{
+		CacheSize:                   defaultCacheSize,
+		BytesPerSync:                512 * units.KiB,
+		WALBytesPerSync:             0, // Default to no background syncing.
+		MemTableStopWritesThreshold: 8,
+		MemTableSize:                defaultCacheSize / 4,
+		MaxOpenFiles:                4096,
+	}
+
+	comparer = pebble.DefaultComparer
+)
+
+type Database struct {
+	lock          sync.RWMutex
+	pebbleDB      *pebble.DB
+	closed        bool
+	openIterators set.Set[*iter]
+}
+
+type Config struct {
+	CacheSize                   int
+	BytesPerSync                int
+	WALBytesPerSync             int // 0 means no background syncing
+	MemTableStopWritesThreshold int
+	MemTableSize                int
+	MaxOpenFiles                int
+	MaxConcurrentCompactions    int
+}
+
+// TODO: Add support for adding a custom logger
+// TODO: Add metrics
+func New(file string, cfg Config, log logging.Logger, _ string, _ prometheus.Registerer) (*Database, error) {
+	opts := &pebble.Options{
+		Cache:                       pebble.NewCache(int64(cfg.CacheSize)),
+		BytesPerSync:                cfg.BytesPerSync,
+		Comparer:                    comparer,
+		WALBytesPerSync:             cfg.WALBytesPerSync,
+		MemTableStopWritesThreshold: cfg.MemTableStopWritesThreshold,
+		MemTableSize:                cfg.MemTableSize,
+		MaxOpenFiles:                cfg.MaxOpenFiles,
+		MaxConcurrentCompactions:    func() int { return cfg.MaxConcurrentCompactions },
+	}
+	opts.Experimental.ReadSamplingMultiplier = -1 // explicitly disable seek compaction
+
+	log.Info("opening pebble")
+	log.Debug("config", zap.Any("config", opts))
+
+	db, err := pebble.Open(file, opts)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Database{
+		pebbleDB:      db,
+		openIterators: set.Set[*iter]{},
+	}, nil
+}
+
+func (db *Database) Close() error {
+	db.lock.Lock()
+	defer db.lock.Unlock()
+
+	if db.closed {
+		return database.ErrClosed
+	}
+
+	db.closed = true
+
+	for iter := range db.openIterators {
+		iter.lock.Lock()
+		iter.release()
+		iter.lock.Unlock()
+	}
+	db.openIterators.Clear()
+
+	return updateError(db.pebbleDB.Close())
+}
+
+func (db *Database) HealthCheck(_ context.Context) (interface{}, error) {
+	db.lock.RLock()
+	defer db.lock.RUnlock()
+
+	if db.closed {
+		return nil, database.ErrClosed
+	}
+	return nil, nil
+}
+
+func (db *Database) Has(key []byte) (bool, error) {
+	db.lock.RLock()
+	defer db.lock.RUnlock()
+
+	if db.closed {
+		return false, database.ErrClosed
+	}
+
+	_, closer, err := db.pebbleDB.Get(key)
```

Comment:
Looks like that []byte will not be valid after closer.Close().  The comments on pebbleDB.get indicate that
> The returned slice will remain valid until the returned Closer is closed.